.packit.sh

```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bash
+
+# Packit's default fix-spec-file often doesn't fetch version string correctly
+# and may also not account for new PRs installing additional files. This script
+# handles any custom processing of the dist-git spec file and gets used by the
+# fix-spec-file action in .packit.yaml
+
+# Get Version from HEAD
+export HEAD_VERSION=$(grep 'var Version = semver.MustParse' version/version.go | cut -d\" -f2 | sed -e 's/-/~/')
+
+# Generate source tarball
+git archive --prefix=podman-$HEAD_VERSION/ -o podman-$HEAD_VERSION.tar.gz HEAD
+
+# RPM Spec modifications
+
+# Fix Version
+sed -i "s/Version: .*/Version: $HEAD_VERSION/" podman.spec
+
+# Fix Release
+sed -i "s/Release: %autorelease/Release: $PACKIT_RPMSPEC_RELEASE%{?dist}/" podman.spec
+
+# Fix Source0
+sed -i "s/%{git0}\/archive\/v%{built_tag_strip}.tar.gz/%{name}-$HEAD_VERSION.tar.gz/" podman.spec
+
+# Fix autosetup
+sed -i "s/%autosetup -Sgit -n %{name}-%{built_tag_strip}/%autosetup -Sgit -n %{name}-$HEAD_VERSION/" podman.spec
+
+# Add any additional installed files not in released versions to %files list
+sed -i "/%{_tmpfilesdir}\/%{name}-docker.conf/a %{_datadir}\/user-tmpfiles.d\/%{name}-docker.conf" podman.spec
```

Comment:
...and here we go, this is what I find scariest. This is a solution to a problem that we have right now, August 24-25 2022, and which will almost certainly change tomorrow or next week and require similar muckery (quite likely performed by some poor schmoe submitting a PR and who will have no clue what this is). My problems are:
- this completely defeats the purpose of packit. We literally want packit to tell us when a PR changes the shipped files, so the specfile can get any necessary changes, right?
- like the above, but, putting this under github-PR control (instead of fedpkg) makes me really queasy. I don't see any major security implications, but it could hide/defer problems.

Isn't this something that should be solved in the specfile itself?
- if this new file should be shipped, can't we ship it?
- if it SHOULD NOT be shipped, can't you do an `rm -f` in the `%install` section?

Reply:
> ...and here we go, this is what I find scariest. This is a solution to a problem that we have right now, August 24-25 2022, and which will almost certainly change tomorrow or next week and require similar muckery (quite likely performed by some poor schmoe submitting a PR and who will have no clue what this is). My problems are:
> 
> * this completely defeats the purpose of packit. We literally want packit to tell us when a PR changes the shipped files, so the specfile can get any necessary changes, right?

Yes, ideally when a packit build fails due to `installed_but_unpackaged` files, it could be handled in spec file itself unless it's a totally new path, for example: this recent case of `/usr/share/user-tmpfiles.d/podman-docker.conf` which was never encountered by podman before.

Committing this file entry to the spec file in dist-git instead of `in-packit` would cause another error: `File not found: /usr/share/user-tmpfiles.d/podman-docker.conf` on any koji builds that would be required before the upstream version containing this file is released.

> * like the above, but, putting this under github-PR control (instead of fedpkg) makes me really queasy. I don't see any major security implications, but it could hide/defer problems.

The file entry here is only meant to make scratch builds happy. The scratch builds aren't meant to be consumed anywhere (maybe they could be consumed by other packit tests, that's for later), but certainly not meant for human consumption.

The canonical fedpkg dist-git copy of podman.spec does not get modified with this script, so we're not really putting any fedora source under github control.

Also, after we commit `/usr/share/user-tmpfiles.d/podman-docker.conf` in fedora's dist-git, even if that entry is not removed from this script, worst-case packit scratch rpm builds will only warn: `File listed twice`, but would not break the build.

> 
> Isn't this something that should be solved in the specfile itself?
> 
> * if this new file should be shipped, can't we ship it?

If it was a new file in an already included path, it could be handle with wildcarded paths, but since this one is a totally new path and since we don't want podman to end up owning entire `/usr/share`,  we add this change separately here, and we can later get rid of it once we build a new release of podman that contains this file.

> * if it SHOULD NOT be shipped, can't you do an `rm -f` in the `%install` section?

While this can be done, I might actually end up forgetting about removing this line and we might have users complain later.

Main takeaway AFAICT, this file addition will be needed in only rare cases

Reply:
Let me try another way to describe my problem with this.

I see packit as something that, if it barfs, means one of two things:
1. Hey, developer, you broke the RPM build. Fix your PR. Or:
2. Hey, specfile maintainer, heads-up! If this PR merges, you will need to update the specfile!

Is that accurate? If it is not, or if I've missed something, then there's no point reading further.

But if the above is correct: (1) will be handled by the developer, yay. (2), however, should absolutely not be handled by the PR developer! S/he likely has no idea what a specfile is, or what packit is, or whow to write a magic `sed` incantation (or where to even find the magic packit.sh file to add it to). This is a heavy burden to place on someone.

Furthermore: adding the `sed` magic here defers the problem: the specfile maintainer will still need to, one day, deal with the problem when build time comes. Which may be far in the future.

As I see it, the best approach is to deal with it in the specfile, immediately (as soon as packit detects the problem). Have the specfile maintainer go to fedpkg and figure out a way to deal with the situation both with and without this PR: 
```
if (packit)   <-- assumes that packit defines an RPM macro that can be tested
...include this file in our ship list
```
The specfile maintainer will still have to re-deal with it at koji-build time, by removing that conditional, but removing existing-code is much much much easier on the specfile maintainer than going and seeing "ah heck, build broke, I remember a PR that added some file a month or two ago, where was that?"

Or:
```
if (such-and-such-file exists)
then include it
```
This is more futureproof, but has a big disadvantage: if the file ever goes away, the specfile will not notice, and will stop shipping it.

There may be other solutions. My point is, I really think if we're going to use packit it should be a commitment on our part to deal with what it reports and do so in a timely manner. Hacking around it in the podman repo itself gives us nothing. (Anti-nothing, actually: I think it harms us).

And I still think the packit.sh has way too much uncomfortable knowledge about the content of the specfile.

Reply:
> Let me try another way to describe my problem with this.
> 
> I see packit as something that, if it barfs, means one of two things:
> 
> 1. Hey, developer, you broke the RPM build. Fix your PR. Or:

Builds can break for many reasons. What if 
- the PR builds on only aarch64 and x86_64 but not on the other arches that fedora supports?
- some vendor dance / dependabot attempt worked well enough to make cirrus happy but would otherwise not build on Fedora

These 2 cases should absolutely be the PR developer's problem. Packit will point these out before merge, which I consider to be a huge advantage. We've had a few releases in the past which failed to build on all arches on Fedora and we noticed that only post-upstream-release.

> 2. Hey, specfile maintainer, heads-up! If this PR merges, you will need to update the specfile!

Yes, this is in most cases easy to do such that the PR developer doesn't need to worry about packit. If it's just a new manpage or something in /usr/share/containers, I can easily include it in the %files section without having to wait for a release or merge such that on the github side of things we can just hit a rerun and things will pass.

Again, this `/usr/share/user-tmpfiles.d/podman-docker.conf` was a totally new dir that podman had never used before, so we would have to wait for a new release which installs that file.

A new file with a totally new path is, I think, not a frequent case which coincided with packit, so guess in a way it's good we saw this problem early on.

Anyway, I can see even infrequent file additions being a bother, so I'm cool with `rm -f` this file until it releases. Worst case there'll be 1 failed build on the koji side after release but calmness on github which is better on second thoughts.

> And I still think the packit.sh has way too much uncomfortable knowledge about the content of the specfile.

So, with the last sed for file addition removed, are you still uncomfortable with the other seds? Actually, we can remove the rest too, but then the scratch build version will be (currently): `4.0.0.rc2`. Packit by default (I think) fetches the version string from the last git tag that was pushed on that release and lists all the commits since that tag in the modified spec file changelog.

If we're cool with ignoring the rpm's Version field, then we can actually delete the entire .packit.sh.

WDYT now?